Fix JMAG Naming Issue

dmnewman3 opened the pull request:
# Issue
Address issue #332 

# Summary
This PR fixes the naming issues present with JMAG 23.0. This solution renames the parts in question affected by the new naming convention.

closes #332

dmnewman3 committed:
add code to specify geometry to avoid naming issues
- main mach_cad file
- tutorial 5 file

elsevers commented:
Thanks @dmnewman3. Does this solution require an earlier version of JMAG (i.e., < v23) to be installed?

dmnewman3 commented:
> Thanks @dmnewman3. Does this solution require an earlier version of JMAG (i.e., < v23) to be installed?

@elsevers This solution does NOT require an earlier version of JMAG to be installed. 

I found the following if we tell JMAG to use the 21.1 geometry editor:
- if an earlier version of JMAG is used than the geometry specified (i.e. v19.1), it will default to that versions geometry editor
- if a newer version of JMAG is used than the geometry specified (i.e. v23.0), it will use the geometry editor specified but does not need to be installed

I did a check with this using Severson02, which only has v21.1 and v23.0 and was able to assign the geometry editor to v22.1 and it functioned properly and imported the regions with the original naming convention.

noguchi-takahiro committed:
Fix Stator ID in imag_2d

noguchi-takahiro commented:
Thank you for updating @dmnewman3. I only have JMAG v22 and v23 installed on my laptop, but the naming issue still occurred.

In my understanding, the reason the material are not set correctly is that the Part name are not set correctly (such as /Region). I found that the groups of Magnet and Coils are not designated as /Region because they were "renamed" upon grouping within eMach. If you ungroup these components in JMAG,
<img width="200" alt="image" src="https://github.com/Severson-Group/eMach/assets/114006024/5eb84db3-0353-44b8-b790-f484e65d5423">

the original Part name, including /Region, will become visible.
<img width="200" alt="image" src="https://github.com/Severson-Group/eMach/assets/114006024/f170ea40-03c8-46d6-9c36-7277a81af56a">
 I think there are two approaches that we can take here.

Approach 1: As indicated the email from Anvar, we can use Part ID instead of the Part name when specifying the materials. While we can proceed with this approach, this means that the Part names will remain as odd ones like StatorCore/Region.

Approach 2: This approach involves directly changing the Part names. I have tried this approach and pushed to this PR. I made sure the eMach tutorial 5 works at least.

By the way, while I was checking the Part ID in eMach, I noticed that the StatorCore ID was assigned incorrectly. When I checked the Part ID List in eMach, it looked like this:

![Screenshot 2024-05-27 220952](https://github.com/Severson-Group/eMach/assets/114006024/8aa5eba1-1c25-4b02-a08c-33e9c7ec030f)
The StatorCore ID is then specified as the 6th value (since number of pole-pair $p=1$ in the image below), i.e., the Stator ID is `17` in eMach.
<img width="500" alt="image" src="https://github.com/Severson-Group/eMach/assets/114006024/beb2d8c0-a9fc-4282-ab47-1cd986bbb9bc">

However, in JMAG, the StatorCore ID was `10`,  which should be 5th value in eMach's Part List.
<img width="400" alt="image" src="https://github.com/Severson-Group/eMach/assets/114006024/ae842180-d075-402e-9d8e-26b8f8434e15">

 I have updated this as well. Unfortunately, this incorrect ID does not seem to be causing the current bug.

What do you think about this approach @dmnewman3?  If you agree with it, could you run the code and check for any errors?

noguchi-takahiro committed:
Uncomment codes to rename parts

dmnewman3 commented:
> Thank you for updating @dmnewman3. I only have JMAG v22 and v23 installed on my laptop, but the naming issue still occurred.
> 
> In my understanding, the reason the material are not set correctly is that the Part name are not set correctly (such as /Region). I found that the groups of Magnet and Coils are not designated as /Region because they were "renamed" upon grouping within eMach. If you ungroup these components in JMAG, <img alt="image" width="200" src="https://private-user-images.githubusercontent.com/114006024/334266067-5eb84db3-0353-44b8-b790-f484e65d5423.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5MDM3MTMsIm5iZiI6MTcxNjkwMzQxMywicGF0aCI6Ii8xMTQwMDYwMjQvMzM0MjY2MDY3LTVlYjg0ZGIzLTAzNTMtNDRiOC1iNzkwLWY0ODRlNjVkNTQyMy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyOFQxMzM2NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZTVmMjU4ZGUyOTIyMjFmNjdhZDZiMDE1ZDQ5ZmNmYWE2MTEwODM2OTFlMzRlYzE2MDBhMjk4OTc5NTc4MWE2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.4nCBP355iDe0XfQMhkZ83OudOgYBs2-B-sE5Ky1yos8">
> 
> the original Part name, including /Region, will become visible. <img alt="image" width="200" src="https://private-user-images.githubusercontent.com/114006024/334266288-f170ea40-03c8-46d6-9c36-7277a81af56a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5MDM3MTMsIm5iZiI6MTcxNjkwMzQxMywicGF0aCI6Ii8xMTQwMDYwMjQvMzM0MjY2Mjg4LWYxNzBlYTQwLTAzYzgtNDZkNi05YzM2LTcyNzdhODFhZjU2YS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyOFQxMzM2NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MTA0MzBiMWU1MDJiNmIzMzhjYTUwOTUxMWFjNGMyODhmY2YwMGY3N2IxNTEyNDBjYzhkMTAyYTYyODY0MWM5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Yd_QwQSIH91Zd3kAMx-fc8cKehrB8zWpyDcwtVJ1PI0"> I think there are two approaches that we can take here.
> 
> Approach 1: As indicated the email from Anvar, we can use Part ID instead of the Part name when specifying the materials. While we can proceed with this approach, this means that the Part names will remain as odd ones like StatorCore/Region.
> 
> Approach 2: This approach involves directly changing the Part names. I have tried this approach and pushed to this PR. I made sure the eMach tutorial 5 works at least.
> 
> By the way, while I was checking the Part ID in eMach, I noticed that the StatorCore ID was assigned incorrectly. When I checked the Part ID List in eMach, it looked like this:
> 
> ![Screenshot 2024-05-27 220952](https://github.com/Severson-Group/eMach/assets/114006024/8aa5eba1-1c25-4b02-a08c-33e9c7ec030f) The StatorCore ID is then specified as the 6th value (since number of pole-pair p=1 in the image below), i.e., the Stator ID is `17` in eMach. <img alt="image" width="500" src="https://github.com/Severson-Group/eMach/assets/114006024/beb2d8c0-a9fc-4282-ab47-1cd986bbb9bc">
> 
> However, in JMAG, the StatorCore ID was `10`, which should be 5th value in eMach's Part List. <img alt="image" width="400" src="https://private-user-images.githubusercontent.com/114006024/334257918-ae842180-d075-402e-9d8e-26b8f8434e15.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5MDM3MTMsIm5iZiI6MTcxNjkwMzQxMywicGF0aCI6Ii8xMTQwMDYwMjQvMzM0MjU3OTE4LWFlODQyMTgwLWQwNzUtNDAyZS05ZDhlLTI2YjhmODQzNGUxNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTI4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUyOFQxMzM2NTNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mNmRmNmUyOWI4YTBhZjE2YTZjNzlhYjYyY2FmNWJmZDQ0YzkzZmY3Y2QyOWFhMWQ4NWQyZTI3YzY2OTFlNTAyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.AHvZq7VSg3q99MymDKAH2zf-gdLvD4r6X0MAzcttEcM">
> 
> I have updated this as well. Unfortunately, this incorrect ID does not seem to be causing the current bug.
> 
> What do you think about this approach @dmnewman3? If you agree with it, could you run the code and check for any errors?

Thank you @noguchi-takahiro, it seems both approaches achieve the same goal. I like the second approach you used (and pushed to the branch) since it involves using the names originally assigned. 

I think it is best we get @elsevers input, since once you and I agree on this, this PR would be ready for his review. I verified the code functions on my end using JMAG23.

elsevers commented on mach_eval/analyzers/electromagnetic/bspm/electrical_analysis/JMAG.py:
```diff
@@ -322,6 +322,7 @@ def draw_jmag_model(
         # open JMAG Geometry Editor
         app.LaunchGeometryEditor()
         geomApp = app.CreateGeometryEditor()
+        geomApp.SetVersion(21.1)
```
Can we remove this line now that we are using the part_ID_list approach?

elsevers commented on mach_cad/tools/jmag/jmag.py:
```diff
@@ -122,6 +122,7 @@ def open(
         self.geometry_editor = self.jd.CreateGeometryEditor(
             True
         )  # creates new geometry or edits the geometry.
+        self.geometry_editor.SetVersion(21.1)
```
Can we remove this line now that we are using the part_ID_list approach?

elsevers commented on mach_eval/analyzers/electromagnetic/bspm/jmag_2d.py:
```diff
@@ -306,6 +306,7 @@ def group(name, id_list):
         partIDRange_Magnet = part_ID_list[2 : int(2 + self.machine_variant.p * 2)]
         # id_sleeve = part_ID_list[int(2 + self.machine_variant.p * 2)]
         id_statorCore = part_ID_list[int(2 + self.machine_variant.p * 2) + 1]
```
Remove extra `id_statorCore` assignment

elsevers reviewed (changes requested):
Nice @dmnewman3 and @noguchi-takahiro! Here's a quick review with a couple requests before a more official L2 review.

After making these changes, please test the code with both versions of JMAG and confirm that it works.

dmnewman3 committed:
make changes based on feedback:
- remove references to version 21.1
- remove extra stator id naming line

dmnewman3 commented:
@elsevers All set, everything still works correctly with the tutorial in question.

noguchi-takahiro reviewed (dismissed):
Thank you for updating @dmnewman3. I also checked It was working in both v22 and v23.

Here is my Level 1 review summary:

Does the code run without error and produce the expected result? Yes
Does the code comply with the [code guidelines](https://emach.readthedocs.io/en/latest/code.html)? Yes
Does the code documentation comply with the [documentation guidelines](https://emach.readthedocs.io/en/latest/documentation.html)? N/A
Is the writing, grammar, and syntax correct and clear? Yes
Is the changeset compliant with the eMach architecture? Yes
Does this review consider whether this physics are accurate? N/A
Is PR approved to Level 2? Yes

@elsevers I approve this PR. Could you review this?

elsevers reviewed (changes requested):
Thanks @dmnewman3 and @noguchi-takahiro.

I see that this PR is only making changes to the bspm analyzer code. We need to test if other analyzers that use JMAG are impacted. Analyzers to test and fix if broken:
1. SynR JMAG 2D FEA analyzer
2. The new flux linkage analyzer in #343 
3. The new BSPM machine constants analyzer in #324
4. eMachPrivate: BIM Transient Analyzer with Two Time-Step Sections

dmnewman3 committed:
fix SynR and flux linkage analyzer for JMAG23

dmnewman3 commented:
> Thanks @dmnewman3 and @noguchi-takahiro.
> 
> I see that this PR is only making changes to the bspm analyzer code. We need to test if other analyzers that use JMAG are impacted. Analyzers to test and fix if broken:
> 
> 1. SynR JMAG 2D FEA analyzer
> 2. The new flux linkage analyzer in [Restructure Flux Linkage and Inductance Analyzers #343](https://github.com/Severson-Group/eMach/pull/343)
> 3. The new BSPM machine constants analyzer in [Add BSPM Machine Constants Analyzer to develop branch #324](https://github.com/Severson-Group/eMach/pull/324)
> 4. eMachPrivate: BIM Transient Analyzer with Two Time-Step Sections

@noguchi-takahiro Can you tackle items 3 and 4 here? I've already done 1 and 2. All that is required is to rename the regions in question. I think you have more experience with these two analyzers than I do. Please push the changes once you've made them and I can re-request a review from @elsevers.

dmnewman3 commented:
> Thanks @dmnewman3 and @noguchi-takahiro.
> 
> I see that this PR is only making changes to the bspm analyzer code. We need to test if other analyzers that use JMAG are impacted. Analyzers to test and fix if broken:
> 
> 1. SynR JMAG 2D FEA analyzer
> 2. The new flux linkage analyzer in [Restructure Flux Linkage and Inductance Analyzers #343](https://github.com/Severson-Group/eMach/pull/343)
> 3. The new BSPM machine constants analyzer in [Add BSPM Machine Constants Analyzer to develop branch #324](https://github.com/Severson-Group/eMach/pull/324)
> 4. eMachPrivate: BIM Transient Analyzer with Two Time-Step Sections

@elsevers All four of these issues have been addressed:
1. Fixed in this PR
2. Fixed and changed in PR #343 
3. Fixed and changed in PR #324 
4. Fixed and changed in eMachPrivate PR [#23](https://github.com/Severson-Group/eMachPrivate/pull/23)

dmnewman3 commented:
@noguchi-takahiro @elsevers Can one of you review this PR so we can get this merged?

elsevers reviewed (approved):
Level 2 review summary: Thanks @dmnewman3. This looks ready to merge to me.

- Does the code comply with the [code guidelines](https://emach.readthedocs.io/en/latest/code.html)? **Yes**
- Does the code documentation comply with the [documentation guidelines](https://emach.readthedocs.io/en/latest/documentation.html)? **Not applicable**
- Is the writing, grammar, and syntax correct and clear? **Not applicable**
- Is the changeset compliant with the eMach architecture? **Yes**
- Are the physics accurate? **Not applicable**
- Did the reviewer change [the release schedule](https://github.com/orgs/Severson-Group/projects/26/views/1) for the issues bing closed? **Yes**, I moved it to a minor revision because it is a bug fix
- Did the reviewer change the branch the PR is being merged into? **No**, this is correctly configured to merge into `develop`
- Level 1 re-review instructions (if revisions are requested): No re-review necessary.
- When your PR is approved, **remember to merge via `Squash and Merge`**. Please ask if you are unsure of how to do this.

dmnewman3 commented:
Thanks @elsevers,
@noguchi-takahiro can you quickly review this and approve? I need 2 approvals in order to merge.
